src/stan/model/indexing/rvalue_cl.hpp

```diff
@@ -0,0 +1,329 @@
+template <typename Expr, typename RowIndex, typename ColIndex,
+          require_rev_kernel_expression_t<Expr>* = nullptr,
+          require_not_same_t<RowIndex, math::matrix_cl<int>>* = nullptr,
+          require_not_same_t<ColIndex, math::matrix_cl<int>>* = nullptr,
+          require_any_not_same_t<RowIndex, ColIndex, index_uni>* = nullptr>
+inline auto rvalue(Expr&& expr, const char* name, const RowIndex row_index,
+                   const ColIndex col_index) {
+  int rows = expr.rows();
+  int cols = expr.cols();
+  internal::index_check(row_index, name, rows);
+  internal::index_check(col_index, name, cols);
+
+  auto res_vari
+      = expr.vi_->index(internal::cl_row_index(row_index, rows, name),
+                        internal::cl_col_index(col_index, expr.cols(), name));
+  return math::var_value<value_type_t<decltype(res_vari)>>(
+      new decltype(res_vari)(std::move(res_vari)));
+}
+
+/**
+ * Index a rev kernel generator expression with two uni indices.
+ *
+ * @tparam Expr type of the expression
+ * @param expr a prim kernel generator expression to index
+ * @param name name of value being indexed (if named, otherwise an empty string)
+ * @param row_index row index
+ * @param col_index column index
+ * @return result of indexing
+ */
+template <typename Expr, require_rev_kernel_expression_t<Expr>* = nullptr>
+inline math::var rvalue(Expr&& expr, const char* name,
+                        const index_uni row_index, const index_uni col_index) {
+  using Val = stan::value_type_t<stan::value_type_t<Expr>>;
+  math::check_range("uni indexing", name, expr.rows(), row_index.n_);
+  math::check_range("uni indexing", name, expr.cols(), col_index.n_);
+  cl::CommandQueue queue = stan::math::opencl_context.queue();
+  Val res;
+  try {
+    cl::Event copy_event;
+    queue.enqueueReadBuffer(
+        expr.val().buffer(), true,
+        sizeof(Val) * (row_index.n_ - 1 + (col_index.n_ - 1) * expr.rows()),
+        sizeof(Val), &res, &expr.val().write_events(), &copy_event);
+    copy_event.wait();
+  } catch (const cl::Error& e) {
+    stan::math::check_opencl_error("uni uni indexing", e);
+  }
+  return math::make_callback_var(
+      res, [expr, row_index, col_index](math::vari res_vari) mutable {
+        block(expr.adj(), row_index.n_, col_index.n_, 1, 1)
+            += math::constant(res_vari.adj(), 1, 1);
+      });
+}
+
+// rev, with multi-index - possible data races in rev, needs special kernel
+/**
+ * Index a rev kernel generator expression with one multi-index.
+ *
+ * @tparam Expr type of the expression
+ * @tparam RowIndex type of index
+ * @param expr a prim kernel generator expression to index
+ * @param name name of value being indexed (if named, otherwise an empty string)
+ * @param row_index index
+ * @return result of indexing
+ */
+template <typename Expr, require_rev_kernel_expression_t<Expr>* = nullptr>
+inline auto rvalue(Expr&& expr, const char* name,
+                   const math::matrix_cl<int>& row_index) {
```

Comment:
> Why not have rvalue() and assign() here that take in std::vector<int> types and call to_matrix_cl() here?

Because if the same index is used in every iteration we only need to copy it to GPU once. I it was done in indexing functions it would be copied every time.

Reply:
So there's two cases I'm thinking about

1.

```stan

data {
  int N;
  int idx[N];
}

parameters {
  vector[N] y;
}

model {
  vector[N] x = exp(y[idx]);
}
```
For that in the compiler I would just make an `idx_cl` that is moved over when the model is constructed (aka what you have written here)

```stan
model {
  int idx2[5] = {1, 2, 3, 4, 5}
  vector[5] x = exp(y[idx2]);
  vector[5] x = exp(y[{1, 2, 3, 4 ,5}]);
}
```

For these I think I'd have to do some odd things to figure out when they need wrapped in `to_matrix_cl()` since they will always be copied since they are constructed in the model block. For these cases it would be easier to just have a version of this that copies a vector over in the function.

Reply:
I don't understand what is odd about this case. `to_matrix_cl()` needs to be called, if you want to call an OpenCL overload of a function and you don't have the data on OpenCL device. This is the same as any other function with OpenCL overloads.

I would really prefer to keep this consistent with all other functions and not add overloads that accept mix of OpenCL and non-OpenCL arguments.

Reply:
I agree with @t4c1 here. The second example is written inefficiently and we dont need to fix those cases. 

Its the same if a user has an expression with only data arguments anywhere outside trans. data. We dont fix those and we dont need to, at least not as a default mode.

Reply:
Okay I think that makes sense. I'm mostly pinging these things as I'm worried about how hard the compiler impl will be. But once I start on it we can sort out whether these things are actually a difficulty or not and do whatever then. I agree this is fine as is